source/common/stats/lazy_init.h

```diff
@@ -0,0 +1,129 @@
+#pragma once
+
+#include "source/common/common/thread.h"
+
+namespace Envoy {
+namespace Stats {
+
+template <typename StatsStructType> class DirectStats;
+template <typename StatsStructType> class LazyInit;
+
+/**
+ * Interface for stats lazy initialization.
+ * To reduce memory and CPU consumption, Envoy can enable the bootstrap config
+ * :ref:`enable_lazyinit_stats
+ * <envoy_v3_api_field_config.bootstrap.v3.Bootstrap.enable_lazyinit_stats>`.
+ * A 'StatsStructType' is only created when any of its field is referenced.
+ * See more context: https://github.com/envoyproxy/envoy/issues/23575
+ */
+template <typename StatsStructType> class LazyCompatibleInterface {
+public:
+  // Helper function to get-or-create and return the StatsStructType object.
+  virtual StatsStructType* instantiate() = 0;
+
+  virtual ~LazyCompatibleInterface() = default;
+};
+
+/**
+ * Lazy-initialization wrapper for StatsStructType, intended for deferred instantiation of a block
+ * of stats that might not be needed in a given Envoy process.
+ *
+ * This class is thread-safe -- instantiations can occur on multiple concurrent threads.
+ * This is used when
+ * :ref:`enable_lazyinit_stats
+ * <envoy_v3_api_field_config.bootstrap.v3.Bootstrap.enable_lazyinit_stats>` is enabled.
+ */
+template <typename StatsStructType>
+class LazyInit : public LazyCompatibleInterface<StatsStructType> {
+public:
+  // Capture the stat names object and the scope with a ctor, that can be used to instantiate a
+  // StatsStructType object later.
+  // Caller should make sure scope and stat_names outlive this object.
+  LazyInit(const typename StatsStructType::StatNameType& stat_names, Stats::ScopeSharedPtr scope)
+      : inited_(Stats::Utility::gaugeFromElements(
+            *scope, {Stats::DynamicName(StatsStructType::typeName()), Stats::DynamicName("inited")},
```

Comment:
oh I missed this before. We need to make a symbolic names for typeName() and "inited". I don't think this is going to be difficult; in fact this is already supported by the macros. For example, in envoy/upstream/upstream.h:635 we have:
```
#define ALL_CLUSTER_TRAFFIC_STATS(COUNTER, GAUGE, HISTOGRAM, TEXT_READOUT, STATNAME)               \
  COUNTER(bind_errors)                                                                             \
  COUNTER(original_dst_host_invalid)                                                               \
  ...
```

That STATNAME macro arg is what you want to use to define a symbolic name here, so you'd have something like:

```
#define ALL_CLUSTER_TRAFFIC_STATS(COUNTER, GAUGE, HISTOGRAM, TEXT_READOUT, STATNAME)               \
  COUNTER(bind_errors)                                                                             \
  COUNTER(original_dst_host_invalid)                                                               \
  STATNAME("initialized")
  STATNAME("traffic_stats")                                                         \  
  ...
```
Of course you'd put the STATNAMEs last in the macro call-site for abc-order consistently.

That will populate the StatName objects in ClusterTrafficStatNames, which is can be used here as `stat_names.initialized_` and `stat_names_.traffic_stats_`

Since our objective here is to save memory it's a bit anachronistic to add some DynamicNames which will replicate the bytes for these names across all the clusters.

Reply:
done